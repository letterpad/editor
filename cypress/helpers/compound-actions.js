// import { cy } from "./serialize";
import { clearEditor, repeatKey, clickXPath } from "./simple-actions";

export async function applyEditorFeatureToLine(xPath) {
  return new Promise(resolve => {
    cy.get('div[contenteditable="true"]').then(() => {
      cy.focused()
        .type("These is a sample line of text{selectAll}")
        .then(() => {
          clickXPath(xPath);
          resolve();
        });
    });
  });
}

export async function applyEditorFeatureToSampleText(xPath) {
  return new Promise(resolve => {
    cy.get('div[contenteditable="true"]').then(() => {
      cy.focused()
        .type("This is a sample text{selectAll}")
        .then(() => {
          clickXPath(xPath);
          resolve();
        });
    });
  });
  // await cy.focused().type("{selectall}{rightarrow}{rightarrow}");
  // // .setSelection("sample");
  // await repeatKey("{leftarrow}", 5); // move 5 left
  // await cy.focused().type("{shift}", { release: false });
  // await repeatKey("{shift}{leftarrow}", 6); // move 6 left
  // await cy.focused().type("{shift}", { release: true });
  // // await cy.focused().type("{selectAll}");
  // await clickXPath(xPath);
}

export async function fillSampleList(cy) {
  await clearEditor(cy);
  await page.keyboard.type("- Item 1");
  await page.keyboard.press("Enter");
  await page.keyboard.type("Item 2");
  await page.keyboard.press("Enter");
  await page.keyboard.press("Tab");
  await page.keyboard.type("Item 2.1");
  await page.keyboard.press("Enter");
  await page.keyboard.type("Item 2.2");
  await page.keyboard.press("Enter");
  await page.keyboard.press("Tab");
  await page.keyboard.type("Item 2.2.1");
  await page.keyboard.press("Enter");
  await page.keyboard.type("Item 2.2.2");
  await page.keyboard.press("Enter");
  await page.keyboard.type("Item 2.2.3");
}
