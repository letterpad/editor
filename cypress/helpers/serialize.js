import { parseFragment, serialize } from "parse5";

function removeData(root) {
  if (root.tagName === "iframe") {
    root.childNodes = [];
  }
  if (root.attrs) {
    root.attrs = root.attrs.filter(
      attr => !attr.name.startsWith("data-") && attr.name !== "class"
    );
  }
  if (root.childNodes) {
    for (const node of root.childNodes) {
      removeData(node);
    }
  }
}

export async function getHtmlContents() {
  return new Promise((resolve, reject) => {
    cy.get("div[contenteditable='true']").then(handle => {
      const fragment = parseFragment(handle[0].innerHTML);
      removeData(fragment);

      resolve(serialize(fragment));
    });
  });
}

export async function textContent(handle) {
  const textContent = await handle.getProperty("textContent");
  return textContent.jsonValue();
}
