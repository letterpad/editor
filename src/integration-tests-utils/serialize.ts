import { ElementHandle } from "puppeteer";
import cheerio from "cheerio";

export type EditorHandle = ElementHandle<Element>;

export async function getHtmlContents(handle: EditorHandle): Promise<string> {
  const context = await handle.executionContext();
  const html = await context.evaluate(node => node.innerHTML, handle);

  const cHtml = cheerio.load(html);
  const allNodes = cHtml("*");
  for (let i = 0; i < allNodes.length; i++) {
    const node = allNodes[i];
    for (let attr in node.attribs) {
      if (attr.startsWith("data-")) {
        allNodes.removeAttr(attr);
      }
    }
  }

  return cHtml.html();
}

export async function textContent(handle: EditorHandle): Promise<any> {
  const textContent = await handle.getProperty("textContent");
  return textContent.jsonValue();
}
