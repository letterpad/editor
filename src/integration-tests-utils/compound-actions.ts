import { EditorHandle } from "./serialize";
import { clearEditor, repeatKey } from "./simple-actions";

export async function applyEditorFeatureToSampleText(
  editorHandle: EditorHandle,
  xPath: string
) {
  await clearEditor(editorHandle);
  await page.keyboard.type("This is a sample text");
  const textHandlers = await page.$x("//span[contains(text(), 'sample')]");
  if (textHandlers.length < 1) {
    throw new Error("Expected at least 1 matching textHandler");
  }

  await repeatKey("ArrowLeft", 5); // move 5 left
  await page.keyboard.down("Shift");
  await repeatKey("ArrowLeft", 6); // move 6 left
  await page.keyboard.up("Shift");
  await page.keyboard.down("Meta");
  await page.keyboard.press("b");
  await page.keyboard.up("Meta");
  await page.waitFor(50);

  const featureHandlers = await page.$x(xPath);
  if (featureHandlers.length < 1) {
    throw new Error(`Feature Button xpath="${xPath}" not found`);
  }

  featureHandlers[0].click();
  await page.waitFor(50);
}
