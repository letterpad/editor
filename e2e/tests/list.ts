import params from "../params";
import { EditorHandle, getHtmlContents } from "../serialize";
import { fillSampleList } from "../compound-actions";
jest.setTimeout(20000);

describe("list", () => {
  let editorHandle: EditorHandle;
  beforeAll(async () => {
    await page.goto(params.testServer);
    await page.waitForXPath("//div[@contenteditable='true']");
    const handle = await page.$('div[contenteditable="true"]');
    editorHandle = handle!;
    await editorHandle.focus();
  });

  test("list works", async () => {
    await fillSampleList(editorHandle);
    expect(await getHtmlContents(editorHandle)).toMatchSnapshot();
  });

  test("shift tab", async () => {
    await fillSampleList(editorHandle);
    await page.keyboard.down("Shift");
    await page.keyboard.press("Tab");
    await page.keyboard.up("Shift");
    expect(await getHtmlContents(editorHandle)).toMatchSnapshot();
  });

  test("backspace", async () => {
    await fillSampleList(editorHandle);
    const currentContent = await getHtmlContents(editorHandle);
    await page.keyboard.press("Enter");
    await page.keyboard.press("Backspace");
    const newContent = await getHtmlContents(editorHandle);
    expect(currentContent).toBe(newContent);
  });

  test("enter should continue in document", async () => {
    await fillSampleList(editorHandle);
    // create a new bullet point
    await page.keyboard.press("Enter");
    // with an empty bullet exit to document
    await page.keyboard.press("Enter");
    // type something
    await page.keyboard.type("This should be part of document.");
    expect(await getHtmlContents(editorHandle)).toMatchSnapshot();
  });
});
