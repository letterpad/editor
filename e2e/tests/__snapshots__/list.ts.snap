// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`list enter should continue in document 1`] = `"<html><head></head><body><ul><li><span><span><span>Item 1</span></span></span></li><li><span><span><span>Item 2</span></span></span></li><ul><li><span><span><span>Item 2.1</span></span></span></li><li><span><span><span>Item 2.2</span></span></span></li><ul><li><span><span><span>Item 2.2.1</span></span></span></li><li><span><span><span>Item 2.2.2</span></span></span></li><li><span><span><span>Item 2.2.3</span></span></span></li></ul></ul></ul><div style=\\"position: relative;\\"><span><span><span>This should be part of document.</span></span></span></div><div style=\\"position: relative;\\"><span><span><span>&#xFEFF;<br></span></span></span></div></body></html>"`;

exports[`list list works 1`] = `"<html><head></head><body><ul><li><span><span><span>Item 1</span></span></span></li><li><span><span><span>Item 2</span></span></span></li><ul><li><span><span><span>Item 2.1</span></span></span></li><li><span><span><span>Item 2.2</span></span></span></li><ul><li><span><span><span>Item 2.2.1</span></span></span></li><li><span><span><span>Item 2.2.2</span></span></span></li><li><span><span><span>Item 2.2.3</span></span></span></li></ul></ul></ul><div style=\\"position: relative;\\"><span><span><span>&#xFEFF;<br></span></span></span></div></body></html>"`;

exports[`list shift tab 1`] = `"<html><head></head><body><ul><li><span><span><span>Item 1</span></span></span></li><li><span><span><span>Item 2</span></span></span></li><ul><li><span><span><span>Item 2.1</span></span></span></li><li><span><span><span>Item 2.2</span></span></span></li><ul><li><span><span><span>Item 2.2.1</span></span></span></li><li><span><span><span>Item 2.2.2</span></span></span></li></ul><li><span><span><span>Item 2.2.3</span></span></span></li></ul></ul><div style=\\"position: relative;\\"><span><span><span>&#xFEFF;<br></span></span></span></div></body></html>"`;
